### PR TITLE
Reduce some allocations

### PR DESCRIPTION
**Pull request by odersky**

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 1]
```diff
@@ -2567,17 +2567,19 @@ class Typer extends Namer
...
         case _ => true
       }
+      var resMatch: Boolean = false
       wtp match {
         case wtp: ExprType =>
           readaptSimplified(tree.withType(wtp.resultType))
-        case wtp: MethodType if wtp.isImplicitMethod && (resultMatch || !functionExpected) =>
+        case wtp: MethodType if wtp.isImplicitMethod &&
+          ({ resMatch = resultMatch; resMatch } || !functionExpected) =>
           if (resultMatch || ctx.mode.is(Mode.ImplicitsEnabled)) adaptNoArgsImplicitMethod(wtp)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]:
```suggestion
          if (resMatch || ctx.mode.is(Mode.ImplicitsEnabled)) adaptNoArgsImplicitMethod(wtp)
```
I think ?

Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Typer.scala]: Oops. yes

**Comment by odersky** [comment]
Some smallish improvements. About 1-2% on compiler, other tests go up to 10%. So, this looks like a worthwhile thing to do.

File: compiler/src/dotty/tools/dotc/ast/Positioned.scala [hunk 2]
```diff
@@ -51,8 +51,12 @@ abstract class Positioned(implicit @constructorOnly src: SourceFile) extends Pro
     else {
       val newpd: this.type =
         if (mySpan.isSynthetic) {
-          if (!mySpan.exists && span.exists)
+          if (!mySpan.exists && span.exists) {
             envelope(source, span.startPos) // fill in children spans
+            () // Note: the `()` is there to prevent some inefficient code from being generated.
+               // Without it we get an allocation of a span here since the result type of the `if`
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/ast/Positioned.scala]: I don't think this changes whether an allocation happens or not but it does make the code more efficient, I've opened an issue to track this that you could refer to in this comment: https://github.com/lampepfl/dotty/issues/5750

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 3]
```diff
@@ -2567,27 +2567,30 @@ class Typer extends Namer
...
       }
+      var resMatch: Boolean = false
       wtp match {
         case wtp: ExprType =>
           readaptSimplified(tree.withType(wtp.resultType))
-        case wtp: MethodType if wtp.isImplicitMethod && (resultMatch || !functionExpected) =>
-          if (resultMatch || ctx.mode.is(Mode.ImplicitsEnabled)) adaptNoArgsImplicitMethod(wtp)
+        case wtp: MethodType if wtp.isImplicitMethod &&
+          ({ resMatch = resultMatch; resMatch } || !functionExpected) =>
+           if (resMatch || ctx.mode.is(Mode.ImplicitsEnabled)) adaptNoArgsImplicitMethod(wtp)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]: oops I added an extra space in my initial suggestion:
```suggestion
          if (resMatch || ctx.mode.is(Mode.ImplicitsEnabled)) adaptNoArgsImplicitMethod(wtp)
```

**Pull request merged**